File: src/core/wasm/pyramid.ts
/** Pyramid parent-tile assembly (mirrors `rust/core/src/abi/pyramid.rs`): halves up to four `size × size` RGBA
 *  child tiles and packs them into one parent tile in a single core call, replacing the per-quadrant
 *  halve+copy loop `TileStore.buildPyramid` (src/storage/tiles.ts) used to run in TS. */
import type { RGBA } from '../../types.ts';
import type { Core } from './core.ts';

/** `children[i]` is quadrant `i` (`dx = i & 1`, `dy = i >> 1`, row-major over the 2×2 grid); `undefined` means
 *  that child tile was never observed and its quadrant comes back zeroed. */
export function assemblePyramidParent(core: Core, children: (Uint8ClampedArray | undefined)[], size: number): RGBA {
  const bytes = size * size * 4;
  const scratchSizes = children.map((c) => (c ? bytes : 0));
  scratchSizes.push(bytes);
  const ptrs = core.scratch(scratchSizes);
  const out = ptrs[4];
  let present = 0;
  for (let i = 0; i < 4; i++) {
    const child = children[i];
    if (child) {
      core.writeBytes(ptrs[i], child);
      present |= 1 << i;
    }
  }
  core.check(
    core.exports.ls_assemble_pyramid_parent(ptrs[0], ptrs[1], ptrs[2], ptrs[3], present, size, out),
    'assemblePyramidParent',
  );
  return { width: size, height: size, data: new Uint8ClampedArray(core.readBytes(out, bytes).buffer) };
}

File: rust/core/src/pyramid.rs
//! Pyramid parent-tile assembly: four child tiles (2×2, possibly with holes where a child tile was never
//! observed) halved and packed into one `size × size` parent tile. The halving itself is `raster::halve_rgba`
//! (alpha-weighted 2:1 reduction); this module only adds the "which quadrant" bookkeeping that used to live in
//! `TileStore.buildPyramid` (src/storage/tiles.ts), one Rust call per parent instead of one per child.

use crate::raster::halve_rgba;

/// Halves each present child (`size × size` RGBA, quadrant order `[dx=0,dy=0]`, `[dx=1,dy=0]`, `[dx=0,dy=1]`,
/// `[dx=1,dy=1]` — i.e. row-major over the 2×2 grid) and copies it into its quadrant of `out` (`size × size`
/// RGBA). A missing child (`None`) leaves its quadrant zeroed, matching the all-zero `Uint8ClampedArray` the TS
/// loop used to start a parent tile from and never touch for an absent child.
pub fn assemble_parent(children: [Option<&[u8]>; 4], size: usize, out: &mut [u8]) {
    let half = size / 2;
    let mut small = vec![0u8; half * half * 4];
    for (i, child) in children.into_iter().enumerate() {
        let (dx, dy) = (i % 2, i / 2);
        small.fill(0);
        if let Some(child) = child {
            halve_rgba(child, size, size, &mut small);
        }
        for r in 0..half {
            let dst = ((dy * half + r) * size + dx * half) * 4;
            out[dst..dst + half * 4].copy_from_slice(&small[r * half * 4..(r + 1) * half * 4]);
        }
    }
}
